src/main/java/org/broadinstitute/hellbender/engine/spark/LocusWalkerSpark.java

```diff
@@ -105,13 +120,11 @@ protected final LIBSDownsamplingInfo getDownsamplingInfo() {
                     new ReferenceMemorySource(bReferenceSource.getValue().getReferenceBases(null, paddedInterval), sequenceDictionary);
             FeatureManager fm = bFeatureManager == null ? null : bFeatureManager.getValue();
 
-            final Set<String> samples = header.getReadGroups().stream()
-                    .map(SAMReadGroupRecord::getSample)
-                    .collect(Collectors.toSet());
-            LocusIteratorByState libs = new LocusIteratorByState(readIterator, downsamplingInfo, false, samples, header, true, false);
-            IntervalOverlappingIterator<AlignmentContext> alignmentContexts = new IntervalOverlappingIterator<>(libs, ImmutableList.of(interval), sequenceDictionary);
-            final Spliterator<AlignmentContext> alignmentContextSpliterator = Spliterators.spliteratorUnknownSize(alignmentContexts, 0);
-            return StreamSupport.stream(alignmentContextSpliterator, false).map(alignmentContext -> {
+            final SAMSequenceDictionary referenceDictionary = reference == null? null : reference.getSequenceDictionary();
+
+            final Iterator<AlignmentContext> alignmentContextIterator = AlignmentContextIteratorFactory.createAlignmentContextIterator(Collections.<SimpleInterval>singletonList(interval), header, readIterator, sequenceDictionary,
```

Comment:
If you make the factory pattern as I suggested, this could be much simpler and you can construct the factory just once keeping the state.